crates/sui-graphql-rpc/schema.graphql

```diff
@@ -0,0 +1,840 @@
+# representation:
+#
+# ([package]::[module]::)?[name](<[param], ...>)?.
+scalar TypeInst
+
+# Move type from a function (parameter or return type) or struct
+# (field) signature.  May contain type parameters, referred to by
+# index.
+scalar TypeSignature
+
+# Filter either by the digest, or the sequence number, or neither, to
+# get the latest checkpoint.
+input CheckpointFilter {
+  digest: Digest
+  sequenceNumber: U64
+}
+
+enum Order { ASC, DESC }
+
+input ObjectQuery {
+  # one of these
+  any: [ObjectQuery]
+  all: [ObjectQuery]
+  not: ObjectQuery
+
+  # (x)or any of these
+
+  # Cascading (type requires module requires package)
+  package: SuiAddress
+  module: String
+  type: String
+
+  owner: SuiAddress
+  objectId: SuiAddress
+  version: U64
+}
+
+input EventFilter {
+  # one of these
+  any: [EventFilter]
+  all: [EventFilter]
+  not: EventFilter
+
+  # (x)or any of these
+  sender: SuiAddress
+  transaction: Digest
+
+  # Cascading (module requires package)
+  emittingPackage: SuiAddress
+  emittingModule: String
+
+  # Cascading (type requires module requires package)
+  eventPackage: SuiAddress
+  eventModule: String
+  eventType: String
+
+  eventField: EventField
+
+  startTimeMs: U64
+  endTimeMs: U64
+}
+
+input EventField {
+  path: String!
+  value: JSON!
+}
+
+input TransactionBlockFilter {
+  # one of these
+  any: [TransactionBlockFilter]
+  all: [TransactionBlockFilter]
+  not: TransactionBlockFilter
+
+  # (x)or any of these
+
+  # Cascading (function requires module requires package)
+  package: SuiAddress
+  module: String
+  function: String
+
+  kind: TransactionBlockKindInput
+  checkpoint: U64
+  fromAddress: SuiAddress
+  toAddress: SuiAddress
+
+  inputObject: SuiAddress
+  changedObject: SuiAddress
+}
+
+type ConsistentRange {
+  first: Checkpoint
+  last: Checkpoint
+}
+
+interface Owner {
+  id: SuiAddress!
+
+  objects(query: ObjectQuery): ObjectQueryConnection
+
+  balance(type: TypeInst!): Balance
+  balances: BalanceConnection
+
+  coins(type: TypeInst): CoinConnection
+
+  stakes: StakeConnection
+
+  defaultNameServiceName: String
+  nameServiceNames: NameServiceNamesConnection
+}
+
+# Returned by Object.owner, where we can't disambiguate between
+# Address and Object.
+type AmbiguousOwner implements Owner {}
+
+type Address implements Owner {
+  transactionBlocks(
+    filter: AddressTransactionBlockFilter,
+  ): TransactionBlockConnection
+}
+
+enum AddressTransactionBlockFilter {
+  SIGN # Transactions this address has signed
+  SENT # Transactions that transferred objects from this address
+  RECV # Transactions that received objects into this address
+  PAID # Transactions that were paid for by this address
+}
+
+type Object implements Owner {
+  owner: AmbiguousOwner
+  version: U64
+  digest: Digest
+
+  previousVersion: U64
+  previousTransactionBlock: TransactionBlock
+  storageRebate: U64
+
+  display: [DisplayEntry]
+
+  asMoveObject: MoveObject
+  asMovePackage: MovePackage
```

Comment:
Downcasting pattern -- @oxade also suggested using a union here, either is fine in my book -- my main motivation for picking this was a thought that it was a pattern that was easier to extend (e.g. if a schema extension wanted to introduce another downcast).

Reply:
This is an interesting approach... I do see how this makes schema extensions easier.

Reply:
Is union any harder to extend?